daemon/cmd/daemon.go

```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	debug.RegisterStatusObject("ongoing-endpoint-creations", d.endpointCreations)
 
 	d.k8sWatcher.RunK8sServiceHandler()
+
+	if option.Config.DNSPolicyUnloadOnShutdown {
+		log.Debug("Registering cleanup function to unload DNS policies")
+
+		// add to pre-cleanup funcs because this needs to run on graceful shutdown, but
+		// before the relevant subystems are being shut down.
+		cleaner.preCleanupFuncs.Add(func() {
+			// Stop k8s watchers
+			log.Info("Stopping k8s service handler")
+			d.k8sWatcher.StopK8sServiceHandler()
+
+			// Iterate over the policy repository and remove L7 DNS part
+			removeL7DNSRules := func(pr policyAPI.Ports) error {
+				portProtocols := pr.GetPortProtocols()
+				if len(portProtocols) != 1 || portProtocols[0].Port != "53" {
+					// not a L7 DNS rule
+					return nil
+				}
```

Comment:
Arguably it's a bit more work, but I think we could iterate all of the rules and look for `len(portProtocols) > 0` + `len(pr.GetPortRule().Rules.DNS) > 0` and always apply the logic. IIRC we do support customising the ports used for DNS and some users remap the DNS ports. So while 'port 53' will catch most cases, we can probably make this a little bit more robust by not assuming this and only relying on the presence of DNS L7 rules in the policies.

Reply:
Good point, I didn't consider that. Changed according to your suggestion.